### PR TITLE
Fix issue in tf.nn.softmax where negative dims could only be -1

### LINKED ISSUE #14916: tf.nn.softmax(input, dim=-1, name=None), argument `dim` cannot take negative index other than -1?
I expect that in the function `tf.nn.softmax(input, dim=-1, name=None)`, the argument `dim` can take dim either positive or negative. However, it seems the only negative index can be taken is -1?

```
import tensorflow as tf

tf.__version__
> '1.4.0'

xx = tf.constant(1, shape=[10, 28, 28, 3], dtype=tf.float32)

tf.nn.softmax(xx, dim=-1)
> <tf.Tensor 'Reshape_1:0' shape=(10, 28, 28, 3) dtype=float32>

tf.nn.softmax(xx, dim=3)
> <tf.Tensor 'Reshape_3:0' shape=(10, 28, 28, 3) dtype=float32>

tf.nn.softmax(xx, dim=-2)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-c0dec6c1fa17> in <module>()
----> 1 tf.nn.softmax(xx, dim=-2)

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/ops/nn_ops.pyc in softmax(logits, dim, name)
   1665       dimension of `logits`.
   1666   """
-> 1667   return _softmax(logits, gen_nn_ops._softmax, dim, name)
   1668 
   1669 

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/ops/nn_ops.pyc in _softmax(logits, compute_op, dim, name)
   1624   # Swap logits' dimension of dim and its last dimension.
   1625   input_rank = array_ops.rank(logits)
-> 1626   logits = _swap_axis(logits, dim, math_ops.subtract(input_rank, 1))
   1627   shape_after_swap = array_ops.shape(logits)
   1628 

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/ops/nn_ops.pyc in _swap_axis(logits, dim_index, last_index, name)
   1596     return array_ops.transpose(logits,
   1597                                array_ops.concat([
-> 1598                                    math_ops.range(dim_index), [last_index],
   1599                                    math_ops.range(dim_index + 1, last_index),
   1600                                    [dim_index]

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/ops/math_ops.pyc in range(start, limit, delta, dtype, name)
   1232       delta = cast(delta, inferred_dtype)
   1233 
-> 1234     return gen_math_ops._range(start, limit, delta, name=name)
   1235 
   1236 

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/ops/gen_math_ops.pyc in _range(start, limit, delta, name)
   3257   if _ctx.in_graph_mode():
   3258     _, _, _op = _op_def_lib._apply_op_helper(
-> 3259         "Range", start=start, limit=limit, delta=delta, name=name)
   3260     _result = _op.outputs[:]
   3261     _inputs_flat = _op.inputs

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.pyc in _apply_op_helper(self, op_type_name, name, **keywords)
    785         op = g.create_op(op_type_name, inputs, output_types, name=scope,
    786                          input_types=input_types, attrs=attr_protos,
--> 787                          op_def=op_def)
    788       return output_structure, op_def.is_stateful, op
    789 

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/framework/ops.pyc in create_op(self, op_type, inputs, dtypes, input_types, name, attrs, op_def, compute_shapes, compute_device)
   2956         op_def=op_def)
   2957     if compute_shapes:
-> 2958       set_shapes_for_outputs(ret)
   2959     self._add_op(ret)
   2960     self._record_op_seen_by_control_dependencies(ret)

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/framework/ops.pyc in set_shapes_for_outputs(op)
   2207       shape_func = _call_cpp_shape_fn_and_require_op
   2208 
-> 2209   shapes = shape_func(op)
   2210   if shapes is None:
   2211     raise RuntimeError(

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/framework/ops.pyc in call_with_requiring(op)
   2157 
   2158   def call_with_requiring(op):
-> 2159     return call_cpp_shape_fn(op, require_shape_fn=True)
   2160 
   2161   _call_cpp_shape_fn_and_require_op = call_with_requiring

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/framework/common_shapes.pyc in call_cpp_shape_fn(op, require_shape_fn)
    625     res = _call_cpp_shape_fn_impl(op, input_tensors_needed,
    626                                   input_tensors_as_shapes_needed,
--> 627                                   require_shape_fn)
    628     if not isinstance(res, dict):
    629       # Handles the case where _call_cpp_shape_fn_impl calls unknown_shape(op).

/home/jetadmin/anaconda2/envs/ygtf/lib/python2.7/site-packages/tensorflow/python/framework/common_shapes.pyc in _call_cpp_shape_fn_impl(op, input_tensors_needed, input_tensors_as_shapes_needed, require_shape_fn)
    689       missing_shape_fn = True
    690     else:
--> 691       raise ValueError(err.message)
    692 
    693   if missing_shape_fn:

ValueError: Requires start <= limit when delta > 0: 0/-2 for 'range' (op: 'Range') with input shapes: [], [], [] and with computed input tensors: input[0] = <0>, input[1] = <-2>, input[2] = <1>.
```

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #14916 where negative dims could only be -1 in tf.nn.softmax.

The issue was that dims=-1 was handled as a case of "last dim" with `is_last_dim = (dim is -1) or (dim == shape.ndims - 1)`, but the generic negative dims were never processed.

This fix adds `dim += shape.ndims` for generic negative dims, and add additional test cases.

This fix fixes #14916.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/ops/nn_ops.py [hunk 1]
```diff
@@ -1626,6 +1626,10 @@ def _swap_axis(logits, dim_index, last_index, name=None):
   # If dim is not the last dimension, we have to do a reshape and transpose so
   # that we can still perform softmax on its last dimension.
 
+  # In case dim is negative (and is not last dimension -1), add shape.ndims
+  if dim < 0:
+    dim += shape.ndims
```
Comment (facaiy) [inline tensorflow/python/ops/nn_ops.py]: Could `dim`  be invalid negative value, say `-5 += 3`?

Comment (yongtang) [reply tensorflow/python/ops/nn_ops.py]: @facaiy In case dim is invalid negative `-5` then a `ValueError` will be thrown.

**Comment by drpngx** [comment]
@facaiy WDYT?

File: tensorflow/python/ops/nn_ops.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/ops/nn_ops.py]: Yup, however `-4 + 3 = -1` would pass, right? Perhaps it's better to check the argument here and raise an invalid value exception early?

**Comment by yongtang** [comment]
@facaiy In the example shown in #14916,
```
xx = tf.constant(1, shape=[10, 28, 28, 3], dtype=tf.float32)
```
so `shape.ndims = 4` (not `3`). In that case, `-4 + 4 = 0` will pass and `-5 + 4 = -1` will fail.

I think the edge case has been covered:
```python
Python 2.7.12 (default, Nov 20 2017, 18:23:56) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> xx = tf.constant(1, shape=[10, 28, 28, 3], dtype=tf.float32)
>>> tf.nn.softmax(xx, dim=-1)
WARNING:tensorflow:From <stdin>:1: calling softmax (from tensorflow.python.ops.nn_ops) with dim is deprecated and will be removed in a future version.
Instructions for updating:
dim is deprecated, use axis instead
<tf.Tensor 'Reshape_1:0' shape=(10, 28, 28, 3) dtype=float32>
>>> tf.nn.softmax(xx, dim=-2)
<tf.Tensor 'transpose_1:0' shape=(10, 28, 28, 3) dtype=float32>
>>> tf.nn.softmax(xx, dim=-3)
<tf.Tensor 'transpose_3:0' shape=(10, 28, 28, 3) dtype=float32>
>>> tf.nn.softmax(xx, dim=-4)
<tf.Tensor 'transpose_5:0' shape=(10, 28, 28, 3) dtype=float32>
>>> tf.nn.softmax(xx, dim=-5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/deprecation.py", line 316, in new_func
    return func(*args, **kwargs)
```

**Comment by facaiy** [comment]
@yongtang Sounds good. Thanks for the clarification!

**Comment by drpngx** [comment]
@aselle WDYT?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 107 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 145 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 160 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 175 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 190 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle, @yuefengz: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle, @yuefengz: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by yongtang** [comment]
@drpngx @facaiy @aselle @yuefengz Any chance to take a look at the PR again? I think the issue raised in review have all been addressed.

**Review by facaiy — DISMISSED** [review]
Thanks, looks good to me.  Would you mind adding a test case for invalid negative dim?

**Comment by facaiy** [comment]
@drpngx @wt-huang I think the change is clear and straightforward. Could you help me to merge it?

**Comment by harshini-gadige** [comment]
> @drpngx @wt-huang I think the change is clear and straightforward. Could you help us to merge it?

@facaiy   -  Let me help you to get this merged as wt-huang is OOO. Do you want me to bypass the kokoro failures ?

**Comment by facaiy** [comment]
@harshini-gadige Thanks. The failures are unrelated. I find that `feedback/copybara` checker is missing, do you know why? If the PR pass in the Google internal repository, then we can merge it.

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/nn_ops.py [hunk 2]
```diff
@@ -1698,6 +1698,10 @@ def _swap_axis(logits, dim_index, last_index, name=None):
   # If dim is not the last dimension, we have to do a transpose so that we can
   # still perform softmax on its last dimension.
 
+  # In case dim is negative (and is not last dimension -1), add shape.ndims
+  if not isinstance(dim, ops.Tensor) and dim < 0:
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow/python/ops/nn_ops.py]: Thanks @alextp, The PR has been updated.

**Comment by yongtang** [comment]
@alextp The PR has been updated, please take a look and let me know if there are any issues.

File: tensorflow/python/ops/nn_ops.py [hunk 3]
```diff
@@ -1698,6 +1698,13 @@ def _swap_axis(logits, dim_index, last_index, name=None):
   # If dim is not the last dimension, we have to do a transpose so that we can
   # still perform softmax on its last dimension.
 
+  # In case dim is negative (and is not last dimension -1), add shape.ndims
+  if not isinstance(dim, ops.Tensor):
+    if dim < 0:
+      dim += shape.ndims
```
Comment (alextp) [inline tensorflow/python/ops/nn_ops.py]: Can you handle unknown shapes here? shape.ndims might be None, in which case you should use array_ops.rank

**Review by alextp — APPROVED** [review]
Thank you!
